### ISSUE TITLE
No errors when passing undefined for required React props

### ISSUE THREAD

**Issue by nirrek** [facebook/flow#2372]
``` js
// @flow
import React from 'react';

class Button extends React.Component {
  props: { title: string };

  render() {
    return <button>{this.props.title}</button>;
  }
}

const ButtonTest = () => (
  <div>
    <Button title={3} />          {/* Type error as expected */}
    <Button title={null} />       {/* Type error as expected */}
    <Button />                    {/* Type error as expected */}
    <Button title={undefined} />  {/* No type error :(       */}
  </div>
);
```

**Expected behavior:**
Passing `undefined` as the `title` prop should lead to a type error because in the prop types `title` is a non-optional, non-nullable string type.

**Observed behavior:**
No type error

**Flow version:**
0.31.0

You can see this [reproduced in the flowtype.org repl here](https://flowtype.org/try/#0PQKgBAAgZgNg9gdzCYAoUYCmBnGBLAOwBcBaAEz2wEMAjGTZNVPAWwAc4AnIsAJUyoBjHlE5wWYAOScBwyQG5UqQTCrZsYAEIBXIkTgEsADyKYCZDfyFEAdAGFxHAmZ4BvVGDBsxbbAC4wVzAiPCJ6AOwiTkIAczAAX0UPMBlzTE4ACgBKQOTPGSJtTkMAHhpdfQIAPlciAAtKG284XxsQsMx4kuByvQMqxU941GHlA0itCvGwAF4wbNmq+eSSigA3KrywEp0+w3b6GddtNKhCTDJ4sGBNz271zazFIA)

If you use a functional component instead of an ES2015 class you do get the expected behavior.

**Comment by gcanti** [comment]
Same with `createElement`

```js
React.createElement(Button, {}) // <= Type error as expected
React.createElement(Button, { title: undefined }) // <= no errors
```

**Comment by guigrpa** [comment]
Might be related: https://github.com/facebook/flow/issues/2781. I understood that the `props` / `defaultProps` type-checking was based on `$Diff<>`

**Comment by asolove** [comment]
This is now fixed in v0.53, which changed the way `Component` prop types get declared.

By just changing the syntax a little, [the original example](https://flow.org/try/#0PTAEAEDMBsHsHcBQBLAtgB1gJwC6gEoCmAhgMZ6RayqgDkWJ5tA3IoqdMQM5egBCAVxw5YAO1CEAHjkKiAJryJkcAOgDC1TKNk4APAG8cyHNEIAuUFxxZkogOYBfADSgAVAD5Q+xKFAN5hFgAFACUXj6+foQ4AljiugBGQiKi7oYAFshcKuhU6NlGJoQOusBJwmLurL4OiLXsYlb8yWIAKoRNALygoaCdnkERunLIAG7uEb66ghXihaad+gDMDqDAnpGbvvrArqCtAJ7ohBJYVFig3BKSx+SEcm7AtZvTLXPGC-qiAtDQq+tbba7fZHE6Bc6XXhSW4yB6uJ6TUCvWZrDaA9FbHZ7Q7HU4Qq7Qwh3OEIl4zFKgeaERYCAKQWz3f4bLGgABysEpoLx2FAZiC6PhzyRwBG40QIWYQA) now throws type errors for all four of the wrong versions in the example.

I think this issue can be closed (/cc @calebmer)
